blackjax/mcmc/proposal.py

```diff
@@ -44,11 +59,21 @@ class Proposal(NamedTuple):
 def proposal_generator(
     kinetic_energy: Callable, divergence_threshold: float
 ) -> Tuple[Callable, Callable]:
-    def new(state: IntegratorState) -> Proposal:
-        energy = -state.logdensity + kinetic_energy(state.momentum)
-        return Proposal(state, energy, 0.0, -np.inf)
+    return proposal_generator_from_energy(
+        energy=lambda state: -state.logdensity + kinetic_energy(state.momentum),
```

Comment:
I agree with this, we want to generalize `proposal_generator` so that it can do acceptance for any target, not just an (independent) potential and kinetic density pair. However, the function here in `proposal.py` should only be the general one and in `hmc.py` the specific `lambda state: -state.logdensity + kinetic_energy(state.momentum)` target should be defined. 

So, I would keep `proposal_generator` as it was, but with any `energy` instead of a `kinetic_energy`, then we can call the specific `energy` (given by the target and kinetic in the case of HMC) inside each specific algorithm's file.

Then, we would need to somehow introduce also a `proposal_generator` for asymmetric proposals, i.e. the `delta_energy` would depend not only on the energy but also on the proposal density. A good example of how it would work is on the `kernel` function of `mcmc.mala.py`, where the acceptance probability of a proposal given by Langevin dynamics needs to include the transition probability since the proposal is not reversible (so asymmetric). In this case, the user would need to pass a function that calculates this probability (what `transition_probability` in `mala.py` does) which would need to be added to the delta energy. 

It might be possible even to use the same `proposal_generator` with an optional `transition_probability` argument which just outputs 0 when not given, but I'm not sure it will work.

Reply:
Hum, recall that the concept of `def hmc_energy(kinetic_energy):
    def energy(state):
        return -state.logdensity + kinetic_energy(state.momentum)
    return energy
` is present not only in hmc, but also in `dynamic_progressive_integration` and `dynamic_recursive_integration` for now I am just extracting that to a function so the fact that is reused is more explicit we can consider further refactoring in the future.

Reply:
> proposal_generator before this PR) assummes that `delta_energy = initial_energy - new_energy` and the initial energy is already calculated, aka there's no way to recalculate it based on the landing state.

Ideally, any algorithm that needs to modify the "inital_energy" based on the proposal would do it before passing it to the `update` function of `proposal_generator`.

> Hum, recall that the concept of def hmc_energy(kinetic_energy): def energy(state): return -state.logdensity + kinetic_energy(state.momentum) return energy is present not only in hmc, but also in dynamic_progressive_integration and dynamic_recursive_integration

Good point, *only* HMC and NUTS are currently using `proposal_generator`. IRMH, RMH, MALA and marginal latent gaussian algorithms all use the same MH step, so they should all be using `proposal_generator`.